ENH: Save BigQuery account credentials in a hidden user folder

parthea opened the pull request:
Closes #41

This PR adds ability for users to choose the location where BigQuery user account credentials will be stored via the `credentials_path` parameter.

This PR also changes the location of the file `bigquery_credentials.dat` that contains the BigQuery user account credentials. Previously this file was saved in the current working directory. The credentials file will now be stored in an application-specific hidden user folder on the operating system. On windows, this path is specified in the `APPDATA` environment variable as defined [here](https://technet.microsoft.com/en-ca/library/cc749104(v=ws.10).aspx) which is typically `'C:\Documents and Settings\username\Application Data\pandas-gbq\'` . Otherwise, credentials will be stored in `'~/.config/pandas-gbq/'`. If the file `bigquery_credentials.dat` is found in the current working directory, the file will be moved to the new location. 

@tswast @jreback Please could you take a look at your earliest convenience?

tswast commented on pandas_gbq/gbq.py:
```diff
@@ -300,14 +315,41 @@ def load_user_account_credentials(self):
+        # Create a pandas_gbq directory in an application-specific hidden
+        # user folder on the operating system.
+        if not os.path.exists(config_path):
+            os.mkdir(config_path)
```
Is this recursive if .config doesn't exist?

parthea replied:
Great catch! `os.mkdir` is not recursive. Fixed.

MaximilianR commented on pandas_gbq/gbq.py:
```diff
@@ -300,14 +315,41 @@ def load_user_account_credentials(self):
+        else:
+            config_path = os.path.join(os.path.expanduser('~'), '.config')
+
+        config_path = os.path.join(config_path, 'pandas_gbq')
```
You know this better than I do, but is there a case for having the default path be the default gcloud path, rather than anything specific to pandas?

parthea replied:
This is a great question! I've given this some thought and I don't feel comfortable writing default credentials into the `gcloud` path in case some users don't want to have their default `gcloud` credentials set (and I'm also worried about colliding with `gcloud`). Alternatively, users could run `gcloud auth application-default login` if they want to configure default credentials and the default `gcloud` credentials should be picked up and won't hit this code path. https://github.com/pydata/pandas-gbq/blob/master/pandas_gbq/gbq.py#L222

@tswast Thoughts?

tswast replied:
+1 to not clobbering `gcloud` credentials. The token we save doesn't have the full Google Cloud Platform scope so that would break other applications that depend on those credentials.

If folks set up application default credentials it won't hit this code anyway.

MaximilianR replied:
> users could run gcloud auth application-default login

> If folks set up application default credentials it won't hit this code anyway.

👍

parthea commented:
I've also made the changes required for pandas.io.gbq in the following branch:
https://github.com/pandas-dev/pandas/compare/master...parthea:sync-with-pandas-gbq

I'll hold off on creating a PR in pandas until this is merged.

parthea commented:
@tswast I've removed the `credentials_path` parameter in favour of the environment variable added in #87. I've rebased the changes with the latest master and everything appears to be working as expected in my local testing. Would you be able to take another look?

tswast commented on docs/source/changelog.rst:
```diff
@@ -1,12 +1,12 @@
 =========
 
-0.2.1 / 2017-??-??
+0.3.0 / 2017-??-??
```
I wonder if we should do an incremental release before `0.3.0` if we plan to have that be the one where we swap to the google-cloud-python libraries? Seems like that PR has stalled and I don't have time to rebase it either.

parthea replied:
Done
